File: main/src/main/java/com/google/android/apps/muzei/util/TickingFloatAnimator.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.util;

import android.animation.TimeInterpolator;
import android.os.SystemClock;
import android.view.animation.AccelerateDecelerateInterpolator;

// Non thread-safe
public class TickingFloatAnimator {
    private float mStartValue = 0;
    private float mCurrentValue;
    private float mEndValue;
    private boolean mRunning = false;
    private long mStartTime;
    private int mDuration = 1000;
    private Runnable mEndCallback;
    private TimeInterpolator mInterpolator = new AccelerateDecelerateInterpolator();

    public static TickingFloatAnimator create() {
        return new TickingFloatAnimator();
    }

    public TickingFloatAnimator from(float startValue) {
        cancel();
        mStartValue = startValue;
        mCurrentValue = startValue;
        return this;
    }

    public TickingFloatAnimator to(float endValue) {
        mEndValue = endValue;
        return this;
    }

    public TickingFloatAnimator withDuration(int duration) {
        mDuration = Math.max(0, duration);
        return this;
    }

    public TickingFloatAnimator withInterpolator(TimeInterpolator interpolator) {
        mInterpolator = interpolator;
        return this;
    }

    public TickingFloatAnimator withEndListener(Runnable listener) {
        mEndCallback = listener;
        return this;
    }

    public void cancel() {
        mRunning = false;
        mEndValue = mCurrentValue;
    }

    public boolean tick() {
        if (!mRunning) {
            return false;
        }

        float t;
        if (mDuration <= 0) {
            t = 1;
        } else {
            t = (float) (SystemClock.elapsedRealtime() - mStartTime) * 1f / mDuration;
            if (t >= 1) {
                t = 1;
            }
        }

        if (t >= 1) {
            // Ended
            mRunning = false;
            mCurrentValue = mEndValue;
            if (mEndCallback != null) {
                mEndCallback.run();
            }
            return false;
        }

        // Still running; compute value
        mCurrentValue = mStartValue + mInterpolator.getInterpolation(t) * (mEndValue - mStartValue);
        mRunning = true;
        return true;
    }

    public void start() {
        mRunning = true;
        mStartValue = mCurrentValue;
        mStartTime = SystemClock.elapsedRealtime();
        tick();
    }

    public boolean isRunning() {
        return mRunning;
    }

    public float currentValue() {
        return mCurrentValue;
    }

    private TickingFloatAnimator() {
    }
}

File: main/src/main/java/com/google/android/apps/muzei/util/TickingFloatAnimator.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.util

import android.animation.TimeInterpolator
import android.os.SystemClock
import android.view.animation.AccelerateDecelerateInterpolator

// Non thread-safe
class TickingFloatAnimator(private val duration: Int) {

    private var startValue: Int = 0
    private var endValue: Int = 0
    private var onEnd: () -> Unit = { }

    private val interpolator: TimeInterpolator = AccelerateDecelerateInterpolator()
    private var startTime: Long = 0
    var currentValue: Float = 0f
    var isRunning = false
        private set

    fun start(startValue: Int = currentValue.toInt(), endValue: Int, onEnd: () -> Unit) {
        this.startValue = startValue
        this.endValue = endValue
        this.onEnd = onEnd
        // Start the TickingFloatAnimator
        isRunning = true
        startTime = SystemClock.elapsedRealtime()
        tick()
    }

    fun tick(): Boolean {
        if (!isRunning) {
            return false
        }

        val t = Math.min((SystemClock.elapsedRealtime() - startTime).toFloat() / duration, 1f)

        isRunning = t < 1f
        currentValue = if (isRunning) {
            startValue + interpolator.getInterpolation(t) * (endValue - startValue)
        } else {
            endValue.toFloat().also { onEnd() }
        }
        return isRunning
    }
}
